File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.py
class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        result = 0
        left = 0
        
        for right in range(1, len(colors)):
            if colors[left] == colors[right]:
                if neededTime[left] > neededTime[right]:
                    result += neededTime[right]                    
                else:
                    result += neededTime[left]
                    left = right
            else:
                left = right
        
        return result
    

# Two Pointers
#Time Complexity: O(n)
#Space Complexity: O(1)

File: 1578-minimum-time-to-make-rope-colorful/1578-minimum-time-to-make-rope-colorful.java
class Solution {
    public int minCost(String colors, int[] neededTime) {
        int result = 0;
        int left = 0;
        
        for(int right = 1; right < colors.length(); right++){
            if(colors.charAt(left) == colors.charAt(right)){
                if(neededTime[left] < neededTime[right]){
                    result += neededTime[left];
                    left = right;
                }
                else{
                    result += neededTime[right];
                }
            }
            
            else{
                left = right;
            }
        }        
        return result;
        
    }
}

// Two Pointers
//Time Complexity: O(n)
//Space Complexity: O(1)